crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -1273,11 +1273,28 @@ impl WritebackCache {
 
     fn clear_state_end_of_epoch_impl(&self, execution_guard: &ExecutionLockWriteGuard<'_>) {
         info!("clearing state at end of epoch");
-        assert!(
-            self.dirty.pending_transaction_writes.is_empty(),
-            "should be empty due to revert_state_update"
-        );
+
+        // Note: there cannot be any concurrent writes to self.dirty while we are in this function,
+        // as all transaction execution is paused.
+        for r in self.dirty.pending_transaction_writes.iter() {
+            let outputs = r.value();
+            for (object_id, object) in outputs.written.iter() {
+                if object.is_package() {
+                    info!("removing non-finalized package from cache: {:?}", object_id);
+                    self.packages.invalidate(object_id);
+                }
+                self.object_by_id_cache.invalidate(object_id);
```

Comment:
So newly created objects are dropped elsewhere? And the package and deleted obj caches do not have the uncertified vs certified separation so needs to be invalidated here?

Reply:
newly created (i.e. dirty) objects are removed when we get to `self.dirty.clear()` below.

Packages are inserted to the package cache when the publish transaction is executed, so we must invalidate the cache here.

`self.object_by_id_cache` can point to entries in the dirty table, so must be cleared.

`self.cached.object_cache` should not, at the current time, contain dirty objects, but I may change that in the future for performance reasons, so I invalidate here just so we don't forget to invalidate if we ever make that change